### PR TITLE
[Payum] Support authorize action

### PR DESCRIPTION
**Pull request by Arminek**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets | #7911 |
| License         | MIT |

### CODE REVIEW

File: src/Sylius/Bundle/PayumBundle/Controller/PayumController.php [hunk 1]
```diff
@@ -123,17 +116,37 @@ public function prepareCaptureAction(Request $request, $tokenValue): Response
...
 
-        $captureToken = $this->getTokenFactory()->createCaptureToken(
-            $payment->getMethod()->getGatewayConfig()->getGatewayName(),
-            $payment,
-            isset($options['route']) ? $options['route'] : null,
-            isset($options['parameters']) ? $options['parameters'] : []
-        );
+        $gatewayConfig = $payment->getMethod()->getGatewayConfig()->getConfig();
+
+        if (isset($gatewayConfig['use_authorize']) && $gatewayConfig['use_authorize'] == true) {
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/PayumBundle/Controller/PayumController.php]: Can we extract this to a private method?

**Comment by pjedrzejewski** [comment]
@Arminek build is failing, can you have a look?

**Comment by pjedrzejewski** [comment]
@psihius @makasim Guys, could you have a quick look?

**Comment by makasim** [comment]
LGTM

File: src/Sylius/Bundle/PayumBundle/Controller/PayumController.php [hunk 2]
```diff
@@ -166,19 +144,45 @@ public function afterCaptureAction(Request $request): Response
...
+    private function provideTokenBasedOnPayment(PaymentInterface $payment, array $redirectOptions): TokenInterface
+    {
+        /** @var GatewayConfigInterface $gatewayConfig */
+        $gatewayConfig = $payment->getMethod()->getGatewayConfig();
+
+        if (isset($gatewayConfig->getConfig()['use_authorize']) && $gatewayConfig->getConfig()['use_authorize'] == true) {
+            $token = $this->getTokenFactory()->createAuthorizeToken(
+                $gatewayConfig->getGatewayName(),
+                $payment,
+                isset($redirectOptions['route'])
```
Comment (pamil) [inline src/Sylius/Bundle/PayumBundle/Controller/PayumController.php]: PHP 7.1 for the rescue - `$redirectOptions['route'] ?? null`

File: src/Sylius/Bundle/PayumBundle/Controller/PayumController.php [hunk 3]
```diff
@@ -166,19 +144,45 @@ public function afterCaptureAction(Request $request): Response
...
+                    ? $redirectOptions['route']
+                    : null,
+                isset($redirectOptions['parameters'])
```
Comment (pamil) [inline src/Sylius/Bundle/PayumBundle/Controller/PayumController.php]: Same here :)

File: src/Sylius/Bundle/PayumBundle/Controller/PayumController.php [hunk 4]
```diff
@@ -166,19 +144,45 @@ public function afterCaptureAction(Request $request): Response
...
     {
         return $this->payum->getHttpRequestVerifier();
     }
+
+    private function provideTokenBasedOnPayment(PaymentInterface $payment, array $redirectOptions): TokenInterface
+    {
+        /** @var GatewayConfigInterface $gatewayConfig */
+        $gatewayConfig = $payment->getMethod()->getGatewayConfig();
+
+        if (isset($gatewayConfig->getConfig()['use_authorize']) && $gatewayConfig->getConfig()['use_authorize'] == true) {
```
Comment (pamil) [inline src/Sylius/Bundle/PayumBundle/Controller/PayumController.php]: `==` -> `===`

File: src/Sylius/Bundle/PayumBundle/spec/Request/GetStatusSpec.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        $this->isRefunded()->shouldReturn(true);
+        $this->getValue()->shouldReturn(PaymentInterface::STATE_REFUNDED);
+    }
+
+    function it_can_be_marked_as_paydout(): void
+    {
+        $this->markPayedout();
+
+        $this->isPayedout()->shouldReturn(true);
+        $this->getValue()->shouldReturn(PaymentInterface::STATE_REFUNDED);
```
Comment (pamil) [inline src/Sylius/Bundle/PayumBundle/spec/Request/GetStatusSpec.php]: Hmm, really?

Comment (pamil) [reply src/Sylius/Bundle/PayumBundle/spec/Request/GetStatusSpec.php]: Ah okay, this one fixes only the authorized state.

**Comment by psihius** [comment]
Don't see any issues with this, but need to pull it and adapt to changes to see if it works properly, that's gonna take time

**Comment by pjedrzejewski** [comment]
Thanks Arek! @psihius We plan to finish the day with RC release, but you could try dev-master already if you find some time. 👍

**Pull request merged**